README.md
# Frequency Key Shifting via CAT on the QCX transceiver

## Introduction 
This software is inspired by [qcx_f](https://github.com/olgierd/qcx_f), which showed that it is possible to transmit FT8 using the [QCX 5W CW transceiver](https://www.qrp-labs.com/qcx.html) with a some creative modifications. It differs from _qcx_f_ in one important aspect: **`qcx_cat_fsk` will work with the QCX's original firmware (1.03+) by using the CAT control interface over serial.** Consequently, one only needs to add the pin header for the serial port (underneath the LCD) on the QCX and a USB-to-serial converter to connect it to a computer. The QCX's earphones audio port is also connected to the computer for receiving signals.

1. `qcx_cat_fsk/audio_decoder.h`: Listens to the audio output from WSJT-X, determines the tone frequency, and accordingly sets VFOB and turns the QCX transmitter on (and off whenever no tones are played.)
2. `qcx_cat_fsk/cat_interface.h`: Implements the Kenwood TS-480-like CAT commands to control the QCX.

This software is currently a proof-of-concept tested with hard-coded values for FT8 on 40m at 7.074MHz. Audio decoding from WSJT-X and frequency shifting via CAT control tested. Despite its narrow CW filter (200Hz) receiving FT8 across portions of the band is also achievable in practice. Confirmed functionality with several QSOs. Next steps are to provide some configurability (e.g., other bands, other FSK-type digital modes), as well as, provide much needed documentation.

## Usage

```
# qcx_cat_fsk --help
usage: qcx_cat_ft8 [--version] [--help] [-l <logfile>]
                   <device name> <frequency>

Controls a QRP Labs QCX CW radio via CAT to transmit a FSK signal decoded from
WSTJ-X output audio. Specify a device name (e.g., /dev/ttyACM0) and the FT8
frequency used in WSJT-X in MHz (e.g., 7.074).

Mandatory arguments to long options are mandatory for short options too.
  -h [ --help ]                      show help
  -v [ --version ]                   show version
  -O [ --offset ] <frequency> (=700) offset in CW frequency in Hz, by default 
                                     set to 700 Hz on the QCX
  -l [ --log ] <logfile>             write log to a specific file, otherwise by
                                     default the log will be written to 
                                     qcx_cat_ft8.log

Report bugs at <https://github.com/jdelacroix/qcx_cat_fsk/issues>.
```

For example, `qcx_cat_ft8 /dev/ttyACM1 7.074` for FT8 with WSJT-X on 40m.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- This software is now versioned, starting at v0.1.0
- `main.cc` now uses `boost::program_options` to provide command-line arguments and options,
  primarily serial port device name, FT8 frequency, log file, and CW offset frequency are
  configurable options rather than hardcoded options

[Unreleased]: https://github.com/jdelacroix/qcx_fsk_cat/compare/v0.1.0...HEAD

src/qcx_cat_fsk/audio_decoder.cc
/*
 * This file is part of qcx_cat_fsk.
 *
 * qcx_cat_fsk is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * qcx_cat_fsk is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with qcx_cat_fsk.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "qcx_cat_fsk/audio_decoder.h"

#include <boost/log/trivial.hpp>
#include <deque>
#include <iomanip>
#include <numeric>
#include <sstream>
#include <stdexcept>

namespace qcx_cat_fsk {

AudioDecoder::AudioDecoder(const std::shared_ptr<CatInterface> &cat_interface,
                           const uint32_t &vfo_a_hz, const uint32_t &vfo_b_hz)
    : vfo_a_hz_(vfo_a_hz), vfo_b_hz_(vfo_b_hz), cat_interface_(cat_interface), is_tx_on_(false) {
  uint32_t audio_sample_rate_hz = 48000;

  static const pa_sample_spec ss = {
      .format = PA_SAMPLE_S16LE, .rate = audio_sample_rate_hz, .channels = 1};

  /* Create the recording stream */
  int error;
  if (!(this->audio_stream_ =
            pa_simple_new(NULL, "qcx_cat_ft8", PA_STREAM_RECORD, NULL, "record",
                          &ss, NULL, NULL, &error))) {
    throw std::runtime_error(pa_strerror(error));
  }
}

void AudioDecoder::Decode() {
  double last_crossing = -1.0, last_frequency = 0;
  std::size_t zero_count = 0,
              zero_count_limit = 48000 / 1000 * 3;  // samples/msec*N

  size_t sample_count = 0, symbol_count = 0;
  std::deque<double> frequencies;

  int16_t last_sample = 0;

  const size_t filter_size = 6;

  int error;

  std::ostringstream vfo_a_cmd_sstream;
  vfo_a_cmd_sstream << "FA" << std::setfill('0') << std::setw(11)
                    << this->vfo_a_hz_
                    << ";";
  std::string vfo_a_cmd = vfo_a_cmd_sstream.str();

  for (;;) {
    std::vector<int16_t> audio_chunk(512, 0);
    /* Record some data ... */
    if (pa_simple_read(this->audio_stream_, audio_chunk.data(),
                       sizeof(int16_t) * audio_chunk.size(), &error) < 0) {
      throw std::runtime_error(pa_strerror(error));
    }
    /* And check for tones. */

    last_crossing = -1.0;

    audio_chunk.insert(audio_chunk.begin(), last_sample);

    for (std::size_t i = 1; i < audio_chunk.size(); ++i) {
      const int16_t sample_a = audio_chunk[i - 1], sample_b = audio_chunk[i];

      // BOOST_LOG_TRIVIAL(info) << "(" << sample_a << ", " << sample_b << ")";

      if (sample_b == 0 && zero_count < zero_count_limit) {  // silence
        ++zero_count;
        if (zero_count == zero_count_limit) {
          try {
            auto reply =
                this->cat_interface_->SendCommand("RX;", false);  // TQ0;
            this->cat_interface_->SendCommand(vfo_a_cmd, false);
            BOOST_LOG_TRIVIAL(info) << "RX; " << reply;
            is_tx_on_ = false;
          } catch (...) {
            // QCX does not reply to this command
          }
          last_frequency = 0;
          frequencies.clear();
          continue;
        }
      }

      if (sample_a < 0 && sample_b >= 0) {  // forward 0-crossing
        zero_count = 0;

        double crossing =
            static_cast<double>(-sample_a) / (sample_b - sample_a) + i;
        if (last_crossing > 0) {
          if (frequencies.size() == filter_size) {
            frequencies.pop_front();
          }
          frequencies.push_back(static_cast<double>(48000) /
                                (crossing - last_crossing));
        }
        last_crossing = crossing;
      }

      if (frequencies.size() == filter_size) {
        double frequency =
            std::accumulate(frequencies.begin(), frequencies.end(), 0.0) /
            frequencies.size();
        if (std::fabs(frequency - last_frequency) > 1) {  // 1Hz
          // frequencies = {frequency};
          frequencies.clear();

          BOOST_LOG_TRIVIAL(info) << "Tone: " << frequency;

          std::stringstream ss;
          ss << "FB" << std::setfill('0') << std::setw(11)
             << (this->vfo_b_hz_ + static_cast<uint32_t>(std::round(frequency)))
             << ";";

          auto cat_command = ss.str();

          try {
            auto reply = this->cat_interface_->SendCommand(cat_command, false);
            BOOST_LOG_TRIVIAL(info) << cat_command << " " << reply;
          } catch (...) {
            // QCX does not reply to this command
          }

          if (is_tx_on_ == false) {
            try {
              auto reply = this->cat_interface_->SendCommand("TQ1;", false);
              BOOST_LOG_TRIVIAL(info) << "TQ1; " << reply;
              is_tx_on_ = true;
            } catch (...) {
              // QCX does not reply to this command
            }
          }

          last_frequency = frequency;
        }
      }
    }

    last_sample = audio_chunk.back();
  }
}

AudioDecoder::~AudioDecoder() {
  if (this->audio_stream_) {
    pa_simple_free(this->audio_stream_);
  }
}

}  // namespace qcx_cat_fsk

src/qcx_cat_fsk/cat_interface.cc
/*
 * This file is part of qcx_cat_fsk.
 *
 * qcx_cat_fsk is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * qcx_cat_fsk is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with qcx_cat_fsk.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "qcx_cat_fsk/cat_interface.h"

#include <boost/lambda/bind.hpp>
#include <boost/log/trivial.hpp>
#include <sstream>

namespace qcx_cat_fsk {

CatInterface::CatInterface(const std::string& qcx_serial_device)
    : serial_port_(io_service_), deadline_timer_(io_service_) {
  this->serial_port_.open(qcx_serial_device);
  this->serial_port_.set_option(boost::asio::serial_port::baud_rate(38400));

  CheckDeadline();
}

std::string CatInterface::SendCommand(const std::string& qcx_cat_command,
                                      const bool wait_for_reply) {
  std::size_t bytes_written = boost::asio::write(
      this->serial_port_, boost::asio::buffer(qcx_cat_command));
  BOOST_LOG_TRIVIAL(info) << "C: [" << qcx_cat_command << ", " << bytes_written
                          << "]";

  if (wait_for_reply == false) {
    return "";
  }

  boost::system::error_code ec = boost::asio::error::would_block;
  boost::asio::streambuf buffer;

  {
    std::lock_guard<std::mutex> lock(this->deadline_timer_mutex_);
    deadline_timer_.expires_from_now(boost::posix_time::seconds(1));
  }
  boost::asio::async_read_until(
      this->serial_port_, buffer, ';',
      (boost::lambda::var(ec) = boost::lambda::_1,
       boost::lambda::var(bytes_written) = boost::lambda::_2));

  do {
    this->io_service_.run_one();
  } while (ec == boost::asio::error::would_block);

  BOOST_LOG_TRIVIAL(info) << "R: [ ," << bytes_written << "]";

  if (ec) {
    throw boost::system::system_error(ec);
  }

  std::stringstream ss;
  ss << &buffer;

  return ss.str();
}

void CatInterface::CheckDeadline() {
  std::lock_guard<std::mutex> lock(this->deadline_timer_mutex_);
  if (this->deadline_timer_.expires_at() <=
      boost::asio::deadline_timer::traits_type::now()) {
    serial_port_.cancel();
    this->deadline_timer_.expires_at(boost::posix_time::pos_infin);
  }
  this->deadline_timer_.async_wait(
      boost::lambda::bind(&CatInterface::CheckDeadline, this));
}

}  // namespace qcx_cat_fsk

include/qcx_cat_fsk/audio_decoder.h
/*
 * This file is part of qcx_cat_fsk.
 *
 * qcx_cat_fsk is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * qcx_cat_fsk is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with qcx_cat_fsk.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once

#include <pulse/error.h>
#include <pulse/simple.h>

#include <memory>
#include <mutex>

#include "qcx_cat_fsk/cat_interface.h"

namespace qcx_cat_fsk {

class AudioDecoder {
 public:
  AudioDecoder(const std::shared_ptr<CatInterface>& cat_interface,
               const uint32_t& vfo_b_hz);
  ~AudioDecoder();

  void Decode();

 private:
  pa_simple* audio_stream_;

  uint32_t vfo_b_hz_;
  std::mutex vfo_b_mutex_;

  std::shared_ptr<CatInterface> cat_interface_;

  bool is_tx_on_;
};
}  // namespace qcx_cat_fsk

include/qcx_cat_fsk/cat_interface.h
/*
 * This file is part of qcx_cat_fsk.
 *
 * qcx_cat_fsk is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * qcx_cat_fsk is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with qcx_cat_fsk.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once

#include <boost/asio.hpp>
#include <boost/asio/deadline_timer.hpp>
#include <boost/asio/serial_port.hpp>
#include <boost/lambda/lambda.hpp>
#include <mutex>
#include <string>

namespace qcx_cat_fsk {

class CatInterface {
 public:
  CatInterface(const std::string& qcx_serial_device);

  std::string SendCommand(const std::string& qcx_cat_command,
                          const bool wait_for_reply = true);

 private:
  void CheckDeadline();

  boost::asio::io_service io_service_;
  boost::asio::serial_port serial_port_;
  boost::asio::deadline_timer deadline_timer_;

  std::mutex deadline_timer_mutex_;
};

}  // namespace qcx_cat_fsk

src/main.cc
/*
 * This file is part of qcx_cat_fsk.
 *
 * qcx_cat_fsk is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * qcx_cat_fsk is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with qcx_cat_fsk.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "qcx_cat_fsk/audio_decoder.h"
#include "qcx_cat_fsk/cat_interface.h"
#include "qcx_cat_fsk/version.h"

#include <boost/log/core.hpp>
#include <boost/log/expressions.hpp>
#include <boost/log/trivial.hpp>
#include <boost/log/utility/setup/file.hpp>
#include <boost/program_options.hpp>

#include <memory>
#include <iomanip>
#include <iostream>
#include <sstream>

int main(int argc, char *argv[]) {

  boost::program_options::positional_options_description arguments;
  arguments.add("device", 1).add("frequency", 1);

  boost::program_options::options_description hidden_options;
  hidden_options.add_options()
    ("device,D", boost::program_options::value<std::string>()->value_name("<name>"), "set device name for connecting to QCX's CAT interface (e.g., /dev/ttyACM0")
    ("frequency,F", boost::program_options::value<double>()->value_name("<frequency>"), "transmit on and decode from this frequency in MHz (e.g., 7.074)");

  boost::program_options::options_description visible_options;
  visible_options.add_options()
    ("help,h", "show help")
    ("version,v", "show version")
    ("offset,O", boost::program_options::value<uint32_t>()->value_name("<frequency>")->default_value(700), "offset in CW frequency in Hz, by default set to 700 Hz on the QCX")
    ("log,l", boost::program_options::value<std::string>()->value_name("<logfile>"), "write log to a specific file, otherwise by default the log will be written to qcx_cat_ft8.log");

  boost::program_options::options_description all_options;
  all_options.add(hidden_options).add(visible_options);

  boost::program_options::variables_map vm;
  try {
    boost::program_options::store(boost::program_options::command_line_parser(argc, argv).options(all_options).positional(arguments).run(), vm);
    boost::program_options::notify(vm);
  } catch(const boost::program_options::error &e) {
    std::cout << "qcx_cat_ft8: " << e.what() << std::endl;
    std::cout << "See 'qcx_cat_ft8 --help'." << std::endl;
    return EXIT_FAILURE;
  }
  
  if (vm.count("help")) {
    std::cout << "usage: qcx_cat_ft8 [--version] [--help] [-l <logfile>]" << std::endl;
    std::cout << "                   <device name> <frequency>" << std::endl;
    std::cout << std::endl;
    std::cout << "Controls a QRP Labs QCX CW radio via CAT to transmit a FSK signal decoded from" << std::endl;
    std::cout << "WSTJ-X output audio. Specify a device name (e.g., /dev/ttyACM0) and the FT8" << std::endl;
    std::cout << "frequency used in WSJT-X in MHz (e.g., 7.074)." << std::endl;
    std::cout << std::endl;
    std::cout << "Mandatory arguments to long options are mandatory for short options too." << std::endl;
    std::cout << visible_options << std::endl;
    std::cout << "Report bugs at <https://github.com/jdelacroix/qcx_cat_fsk/issues>." << std::endl;
    return EXIT_SUCCESS;
  }

  if (vm.count("version")) {
    std::cout << "qcx_cat_ft8 version " << VERSION_MAJOR << "." << VERSION_MINOR << "." << VERSION_PATCH << std::endl;
    return EXIT_SUCCESS;
  }

  if (vm.count("device") == 0 || vm.count("frequency") == 0) {
    std::cout << "qcx_cat_ft8: missing <device name> or <frequency> argument." << std::endl; 
    std::cout << "See 'qcx_cat_ft8 --help'." << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count("log")) {
    boost::log::add_file_log(boost::log::keywords::file_name = vm["log"].as<std::string>(),
                           boost::log::keywords::auto_flush = true);
  } else {
    boost::log::add_file_log(boost::log::keywords::file_name = "qcx_cat_ft8.log",
                           boost::log::keywords::auto_flush = true);
  }

  
  boost::log::core::get()->set_filter(boost::log::trivial::severity >=
                                      boost::log::trivial::info);

  BOOST_LOG_TRIVIAL(info) << "!log initialized";

  // FIXME: hard coded serial port device
  std::shared_ptr<qcx_cat_fsk::CatInterface> qcx;
  try {
    qcx = std::make_shared<qcx_cat_fsk::CatInterface>(vm["device"].as<std::string>());
  } catch (const boost::system::system_error &e) {
    std::cout << "qcx_cat_ft8: unable to connect to QCX via " << vm["device"].as<std::string>() << std::endl;
    // return EXIT_FAILURE;
  }

  uint32_t vfo_b_hz = static_cast<uint32_t>(vm["frequency"].as<double>()*1e6);
  uint32_t vfo_a_hz = vfo_b_hz + vm["offset"].as<uint32_t>();

  std::ostringstream vfo_a_cmd_sstream;
  vfo_a_cmd_sstream << "FA" << std::setfill('0') << std::setw(11)
                    << vfo_a_hz
                    << ";";
  std::string vfo_a_cmd = vfo_a_cmd_sstream.str();

  try {
    auto reply = qcx->SendCommand("FT2;");
    BOOST_LOG_TRIVIAL(info) << "FT2; " << reply;
  } catch (...) {
    // no reply from QCX on this command
  }

  try {
    auto reply = qcx->SendCommand(vfo_a_cmd); // "FA00007074700;");
    BOOST_LOG_TRIVIAL(info) << vfo_a_cmd << " " << reply;
  } catch (...) {
    // no reply from QCX on this command
  }

  // FIXME: hard coded FT8 frequency
  qcx_cat_fsk::AudioDecoder ft8_decoder(qcx, vfo_a_hz, vfo_b_hz);

  ft8_decoder.Decode();
}
